### PR TITLE
Discord Integration! Sync your roles between your NamelessMC site and your Discord Guild.

### PR DESCRIPTION
**Pull request by Aberdeener**
Will require the Nameless Link bot in your discord guild, which can be easily invited.
Once the bot is in your guild, it will ask the owner to send it a pm which contains the site API URL.
If the url is valid, the bot will listen for role changes and update roles accordingly.
In order for a role change to happen, the affected user needs to have their discord user id set on their nameless site. this can be done by adding it in their profile settings, and then confirming it by PMing the Nameless Link bot their site username.

Still a lil' to do, but making a draft PR so I can add status update comments :p 

Major goals:
- ~~Discord integration option on site (if not enabled, roles will not be synced)~~ [DONE]
- ~~Discord User ID linking, with extra verification step mentioned above~~ [DONE]
- ~~Set Discord Role ID for each group~~ [DONE]
- ~~Website -> Discord role syncing~~ [DONE]
- ~~Discord -> Website role syncing~~ [DONE]

Minor goals:
- ~~Error handling when Nameless Link fails to complete a task (verification of Discord ID, etc)~~ [DONE]
- ~~Give the user their website role in the website's guild when they verify their Discord ID~~ [DONE]

Status: Good to go! Awaiting a final review.

### CODE REVIEW

File: modules/Core/pages/panel/discord.php [hunk 1]
```diff
@@ -0,0 +1,136 @@
+<?php
+/*
+ *	Made by Samerton
```
Comment (Derkades) [inline modules/Core/pages/panel/discord.php:3]: Is it? :P

Comment (Aberdeener) [reply modules/Core/pages/panel/discord.php:3]: 👀 I thought that meant NamelessMC as a whole, not the page

**Comment by HeavenCrafter** [comment]
Awww why you gotta release this now, right when I can't afford to host a forum anymore. xD

File: modules/Core/pages/panel/discord.php (hunk 1 above)
Comment (Derkades) [reply modules/Core/pages/panel/discord.php:3]: Not sure to be honest, I never understood the point of header comments anyway

**Comment by Aberdeener** [comment]
Update: The [Nameless-Link](https://github.com/NamelessMC/Nameless-Link) bot is now open source.

File: modules/Core/pages/panel/users_edit.php [hunk 2]
```diff
@@ -191,7 +191,44 @@
...
+							if ($user_query->discord_id != null && $user_query->discord_id != 010) {
+								$group_discord_id = $queries->getWhere('groups', array('id', '=', $group))[0]->discord_role_id;
+								$old_group_discord_id = $queries->getWhere('groups', array('id', '=', $user_query->group_id))[0]->discord_role_id;
+
+								$api_key = $queries->getWhere('settings', array('name', '=', 'mc_api_key'))[0]->value;
+								$api_url = rtrim(Util::getSelfURL(), '/') . rtrim(URL::build('/api/v2/' . Output::getClean($api_key), '', 'non-friendly'), '/');
+								
+								// The bot can handle null roles, but it is better to deal with it here
+								if ($group_discord_id == null && $old_group_discord_id != null) {
+									$full_url = BOT_URL . '/roleChange?id=' . $user_query->discord_id . '&guild_id=' . $queries->getWhere('settings', array('name', '=', 'discord'))[0]->value . '&oldRole=' . $old_group_discord_id;
```
Comment (samerton) [inline modules/Core/pages/panel/users_edit.php:210]: Might need to split this line + 212/214 up

File: modules/Core/pages/user/settings.php [hunk 3]
```diff
@@ -705,6 +781,23 @@
...
+			'DISCORD_ID_VALUE' => $user->data()->discord_id,
+		));
+	} else {
+		$smarty->assign(array(
+			'NOT_LINKED' => $language->get('user', 'not_linked'),
+		));
+		if ($user->data()->discord_id == 010) {
+			$smarty->assign(array(
+				'PENDING_LINK' => $language->get('user', 'pending_link'),
+				''
```
Comment (samerton) [inline modules/Core/pages/user/settings.php:796]: Can this second element be removed?

File: modules/Core/pages/validate.php [hunk 4]
```diff
@@ -45,6 +45,22 @@
 			'language' => $language
 		));
 
+		// Discord integration is enabled
+		if ($queries->getWhere('settings', array('name', '=', 'discord_integration'))[0]->value == '1') {
```
Comment (samerton) [inline modules/Core/pages/validate.php:49]: Might need to split this line up

File: modules/Core/pages/validate.php [hunk 5]
```diff
@@ -45,6 +45,22 @@
...
+			// They have a valid discord Id
+			if ($user->data()->discord_id != null && $user->data()->discord_id != 010) {
+				$group_discord_id = $queries->getWhere('groups', array('id', '=', $user->data()->group_id))[0]->discord_role_id;
```
Comment (samerton) [inline modules/Core/pages/validate.php:52]: Might need to split this line up + line 56 + line 58

File: core/init.php [hunk 6]
```diff
@@ -66,6 +66,9 @@
     // Friendly URLs?
     define('FRIENDLY_URLS', Config::get('core/friendly'));
 
+    // Define discord bot url for easy change if they choose to selfhost
+    define('BOT_URL', 'http://bot.tadhgboyle.dev:8001');
```
[the remark this answers is not in the document]
Comment (Aberdeener) [reply core/init.php:70]: Yeah, good point.
I'll DM you about getting a subdomain of namelessmc.com redirected to my dedi, unless you're okay with me having access to a portion of the namelessmc.com infrastructore (for updating and checking logs etc)

[the remark this answers is not in the document]
Comment (Aberdeener) [reply core/init.php:70]: Added option in StaffCP

File: core/installation/views/installer.view.php [hunk 7]
```diff
@@ -620,7 +620,7 @@
...
...dmincp.core.reactions":1,"admincp.core.registration":1,"admincp.core.social_media":1,"admincp.core.terms":1,"admincp.errors":1,"admincp.integrations":1,"admincp.discord":1,"admincp.minecraft":1,"admincp.minecraft.authme":1,"admincp.minecraft.verification":1,"admincp.minecraft.servers":1,"admincp.minecraft.query_errors":1,"admincp.minecraft.banners":1,"admincp.modules":1,"admincp.pages":1,"admincp.pages.metadata":1,"admincp.security":1,"admincp.security.acp_logins":1,"admincp.security.template":1,"admincp.sitemap":1,"admincp.styles":1,"admincp.styles.panel_templates":1,"admincp.styles.templates":1,"admincp.styles.templates.edit":1,"admincp.styles.images":1,"admincp.update":1,"admincp.users":1,"admincp.users.edit":1,"admincp.groups":1,"admincp.groups.self":1,"admincp.widgets":1,"modcp.ip_lookup":1,"modcp.punishments":1,"modcp.punishments.warn":1,"modcp.punishments.ban":1,"modcp.punishments.banip":1,"modcp.punishments.revoke":1,"modcp.reports":1,"modcp.profile_banner_reset":1,"usercp.messaging":1,"usercp.signature":1,"admincp.forums":1,"usercp.private_profile":1,"usercp.nickname":1,"usercp.profile_banner":1,"profile.private.bypass":1, "admincp.security.all":1,"admincp.core.hooks":1}',
```
[the remark this answers is not in the document]
Comment (Aberdeener) [reply core/installation/views/installer.view.php:623]: Solved

File: custom/languages/EnglishUK/admin.php [hunk 8]
```diff
@@ -660,4 +660,15 @@
...
+	// Discord
+	'discord' => 'Discord',
+	'enable_discord_integration' => 'Enable Discord integration?',
+	'discord_role_id' => 'Discord Role ID',
+	'discord_role_id_numeric' => 'Discord Role ID must be numeric.',
+	'discord_role_id_length' => 'Discord Role ID must be 18 digits long.',
+	'discord_guild_id_updated' => 'Discord Server ID was updated successfully.',
+	'discord_guild_id_required' => 'Please enter your Discord Server ID for Discord integration to work.',
+	'discord_invalid_api_url' => 'Looks like your API URL has changed. Please get the guild owner to update the URL with the Nameless Link bot.'
+	// TODO: All language files
```
[the remark this answers is not in the document]
Comment (Aberdeener) [reply custom/languages/EnglishUK/admin.php:673]: Hahaha I'll try :p

File: core/installation/views/installer.view.php [hunk 9]
```diff
@@ -897,7 +897,12 @@
             $queries->create('settings', array(
                 'name' => 'mc_integration',
                 'value' => 1
-            ));
+			));
+
```
[the remark this answers is not in the document]
Comment (Aberdeener) [reply core/installation/views/installer.view.php:901]: Done

File: modules/Core/hooks/ValidateHook.php [hunk 10]
```diff
@@ -15,5 +15,24 @@ public static function validatePromote($params = array()){
...
+            // They have a valid discord Id
+            if ($user_query->discord_id != null && $user_query->discord_id != 010) {
+                $group_discord_id = $queries->getWhere('groups', array('id', '=', VALIDATED_DEFAULT))[0]->discord_role_id;
+
+                if ($group_discord_id != null) {
+                    $bot_url = BOT_URL;
+                    $api_key = $queries->getWhere('settings', array('name', '=', 'mc_api_key'))[0]->value;
+                    $api_url = rtrim(Util::getSelfURL(), '/') . rtrim(URL::build('/api/v2/' . Output::getClean($api_key), '', 'non-friendly'), '/');
+                    $full_url = $bot_url . '/roleChange?id=' . $user_query->discord_id . '&guild_id=' . $queries->getWhere('settings', array('name', '=', 'discord'))[0]->value . '&ole=' . $group_discord_id;
+                    file_get_contents($full_url);
```
[the remark this answers is not in the document]
Comment (Aberdeener) [reply modules/Core/hooks/ValidateHook.php:33]: Yes, was added yesterday and not put in this code by accident. 
The API URL is passed in the $full_url variable as a parameter, and then the bot makes sure it matches the API URL stored in its database with the guild_id

**Comment by Aberdeener** [comment]
@samerton I realize 8b7eeffcadbce864452cc5a2ee3b894eb170de10 was kind of stupid - sorry

**Pull request merged**